File: chaerim/Programmers/Level_2/Lv2_뒤에_있는_큰_수_찾기.py
def solution(numbers):
    answer = [-1] * len(numbers)
    stack = []

    for i, number in enumerate(numbers):
        while stack and numbers[stack[-1]] < number:
            last = stack.pop()
            answer[last] = number
        stack.append(i)

    return answer

'''시간 초과
def solution(numbers):
    answer = []

    number_max = max(numbers)

    for number in range(len(numbers)):
        larger_number = -1

        if numbers[number] < number_max:
            for i in range(number+1, len(numbers)):
                if numbers[i] > numbers[number]:
                    larger_number = numbers[i]
                    break

        answer.append(larger_number)

    return answer
'''

File: seongwon/Programmers/Level_2/Lv2_뒤에_있는_큰_수_찾기.cpp
#include <vector>
#include <stack>
using namespace std;

vector<int> solution(vector<int> numbers) {
    int n = numbers.size();
    vector<int> answer(n, -1);
    stack<int> s;
    for(int i = 0; i < n; i++){
        while(!s.empty() && numbers[s.top()] < numbers[i]){
            answer[s.top()] = numbers[i];
            s.pop();
        }
        s.push(i);
    }
    return answer;
}
